tests/integration2/pilot/sidecar_api_test.go

```diff
@@ -0,0 +1,134 @@
+package pilot
+
+import (
+	"fmt"
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/api/components"
+	"istio.io/istio/pkg/test/framework/api/descriptors"
+	"istio.io/istio/pkg/test/framework/api/ids"
+	"istio.io/istio/pkg/test/framework/api/lifecycle"
+	"istio.io/istio/pkg/test/framework/runtime/components/environment/native"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/structpath"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func TestSidecarListeners(t *testing.T) {
+	// Call Requires to explicitly initialize dependencies that the test needs.
+	ctx := framework.GetContext(t)
+	scopes.Framework.SetOutputLevel(log.DebugLevel)
+
+	// TODO: Limit to Native environment until the Kubernetes environment is supported in the Galley
+	// component
+	ctx.RequireOrSkip(t, lifecycle.Test, &descriptors.NativeEnvironment)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Galley)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Pilot)
+
+	// Load the canonical dataset into Galley and by implication Pilot
+	gal := components.GetGalley(ctx, t)
+	env, err := native.GetEnvironment(ctx)
+	components.GetMixer(ctx, t)
+	s := strings.Split(env.Mesh.MixerCheckServer, ":")[1]
+	mixerPort, _ := strconv.ParseUint(s, 10, 32)
+
+	path, err := filepath.Abs("../../testdata/config")
+	if err != nil {
+		t.Fatalf("No such directory: %v", err)
+	}
+	gal.ApplyConfigDir(path)
+
+	pilot := components.GetPilot(ctx, t)
+
+	// Simulate proxy identity of egress gateway ...
+	nodeId := &model.Proxy{
+		ClusterID:       "integration-test",
+		Type:            model.SidecarProxy,
+		IPAddresses:     []string {"10.2.0.1"},
+		ID:              "app3.testns",
+		DNSDomain:       "testns.cluster.local",
+	}
+
+	// ... and get listeners from Pilot for that proxy
+	req := &xdsapi.DiscoveryRequest{
+		Node: &core.Node{
+			Id:  nodeId.ServiceNode(),
+		},
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Listener",
+	}
+	resp, err := pilot.CallDiscovery(req)
+	if err != nil {
+		t.Fatalf("discovery failed: %v", err)
+	}
+
+	response, _ := structpath.New(resp)
+	t.Run("validate-mixer-listener", func(t *testing.T) {
+		// Note use %f when matching numerics as all parsed JSON numbers are treated as float
+		mixerListener := response.FindStruct(t, fmt.Sprintf("{.resources[?(@.address.socketAddress.portValue==%v)]}", mixerPort))
```

Comment:
I think this is pretty close to being a fairly self-contained DSL.

Just from an aesthetics/readability perspective, the fluent approach might make things much easier to read. E.g.:
```go
structpath.
  AssertThat(t, response).
  Select("{.resources[?(@.address.socketAddress.portValue==%v)]}", mixerPort).
  Equals("{.filterChains[0].filters[0].config.tracing.client_sampling.value}", 100).
  Matches("{.address.socketAddress.address}", "0.0.0.0").
  IsTrue("{.filterChains[0].filters[0].config.generate_request_id}").
  Exists("{.filterChains[0].filters[0].config.rds.config_source.ads}").
  NotExists("{.useOriginalDst}")  
```

Reply:
Done